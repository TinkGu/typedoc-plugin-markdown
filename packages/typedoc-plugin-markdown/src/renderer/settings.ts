import { DeclarationReflection, ProjectReflection } from 'typedoc';

import { PluginOptions } from './types';

let _project: ProjectReflection;
let _options: PluginOptions;
let _activeUrl: string;
let _activeReflection: DeclarationReflection;

export default {
  get project() {
    return _project;
  },

  set project(project: ProjectReflection) {
    _project = project;
  },

  get options() {
    return _options;
  },

  set options(options: PluginOptions) {
    _options = options;
  },

  get activeUrl() {
    return _activeUrl;
  },

  set activeUrl(url) {
    _activeUrl = url;
  },

  get activeReflection() {
    return _activeReflection;
  },

  set activeReflection(reflection) {
    _activeReflection = reflection;
  },

  get globalsFile() {
    return 'modules.md';
  },

  get globalsName() {
    return _options.entryPoints && _options.entryPoints.length > 1
      ? 'Modules'
      : 'Exports';
  },
};
