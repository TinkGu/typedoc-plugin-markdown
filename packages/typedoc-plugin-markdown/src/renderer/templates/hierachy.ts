import { DeclarationHierarchy } from 'typedoc';
import { spaces } from '../tools/utils';
import { typeTemplate } from './type';

export function hierarchyTemplate(model: DeclarationHierarchy) {
  return hierarchy(model);
}

function hierarchy(model: DeclarationHierarchy, level = 0) {
  const md: string[] = [];
  const symbol = level > 0 ? getSymbol(level) : '-';
  model.types.forEach((hierarchyType) => {
    if (model.isTarget) {
      md.push(`${symbol} **\`${hierarchyType}\`**`);
    } else {
      md.push(`${symbol} ${typeTemplate(hierarchyType)}`);
    }
  });
  if (model.next) {
    md.push(hierarchy(model.next, level + 1));
  }
  return md.join('\n\n');
}

function getSymbol(level: number) {
  return spaces(2) + [...Array(level)].map(() => 'â†³').join('');
}
