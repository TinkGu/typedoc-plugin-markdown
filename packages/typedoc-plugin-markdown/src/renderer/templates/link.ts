import * as path from 'path';

import settings from '../settings';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, activeUrl } = settings;

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
