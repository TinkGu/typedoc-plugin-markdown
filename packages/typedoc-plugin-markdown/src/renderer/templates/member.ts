import { DeclarationReflection, ReferenceReflection } from 'typedoc';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { escapeChars } from '../tools/utils';
import { declarationTemplate } from './declaration';
import { signatureTemplate } from './signature';

export function memberTemplate(
  model: DeclarationReflection | ReferenceReflection,
) {
  const options = settings.options;
  const md = new MarkdownBuilder();

  md.add(
    heading(
      3,
      options.namedAnchors
        ? `<a id="${model.anchor}" name="${model.anchor}"></a> `
        : '' + escapeChars(model.name),
    ),
  );

  if (model.signatures) {
    model.signatures.forEach((signature) => {
      md.add(signatureTemplate(signature));
    });
  } else {
    if (model.hasGetterOrSetter()) {
      if (model.getSignature) {
        md.add(signatureTemplate(model.getSignature, false, 'get'));
      }
      if (model.setSignature) {
        md.add(signatureTemplate(model.setSignature, false, 'set'));
      }
    } else {
      if (model instanceof ReferenceReflection && model.isReference) {
        const deep = model.tryGetTargetReflectionDeep();
        if (deep) {
          md.add(`Re-exports ${deep.name}`);
        }
      }

      if (model instanceof DeclarationReflection) {
        md.add(declarationTemplate(model));
      }
    }
  }

  return md.print();
}
