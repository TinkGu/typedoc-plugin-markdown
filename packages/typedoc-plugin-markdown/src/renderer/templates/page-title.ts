import { ParameterReflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import settings from '../settings';

export function pageTitleTemplate(page: PageEvent, shouldEscape = true) {
  const title: string[] = [''];
  if (page.model && page.model.kindString && page.url !== page.project.url) {
    title.push(`${page.model.kindString}: `);
  }
  if (page.url === page.project.url) {
    title.push(settings.options.indexTitle || page.project.name);
  } else {
    title.push(shouldEscape ? escape(page.model.name) : page.model.name);
    if (page.model.typeParameters) {
      const typeParameters = page.model.typeParameters
        .map((typeParameter: ParameterReflection) => typeParameter.name)
        .join(', ');
      title.push(`<${typeParameters}${shouldEscape ? '\\>' : '>'}`);
    }
  }
  return title.join('');
}
