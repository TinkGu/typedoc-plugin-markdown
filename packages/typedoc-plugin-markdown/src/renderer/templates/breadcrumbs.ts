import { Reflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent) {
  const { activeUrl, globalsName, globalsFile, options } = settings;

  const md = new MarkdownBuilder();

  md.add(
    page.url === options.entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, options.entryDocument),
  );

  if (!options.readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
