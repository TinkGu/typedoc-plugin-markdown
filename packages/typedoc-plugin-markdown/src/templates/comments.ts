import { Comment } from 'typedoc/dist/lib/models';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { backTicks, bold, link } from '../tools/elements';
import { linkTemplate } from './link';

const URL_PREFIX = /^(http|ftp)s?:\/\//;
const BRACKETS = /\[\[([^\]]+)\]\]/g;
const INLINE_TAG =
  /(?:\[(.+?)\])?\{@(link|linkcode|linkplain)\s+((?:.|\n)+?)\}/gi;

export function commentsTemplate(model: Comment) {
  const md = new MarkdownBuilder();

  if (model.hasVisibleComponent()) {
    if (model.shortText) {
      md.add(comment(model.shortText));
    }
    if (model.text) {
      md.add(comment(model.text));
    }
    if (model.tags) {
      const tags = model.tags.map(
        (tag) =>
          bold(backTicks(tag.tagName)) +
          (tag.text
            ? comment((tag.text.startsWith('\n') ? '' : ' ') + tag.text)
            : ''),
      );
      md.add(tags.join('\n\n'));
    }
  }
  return md.print();
}

export function comment(text: string) {
  return replaceInlineTags(replaceBrackets(text));
}

function replaceBrackets(text: string) {
  return text.replace(BRACKETS, (match: string, content: string): string => {
    const split = splitLinkText(content);
    return buildLink(match, split.target, split.caption);
  });
}

function replaceInlineTags(text: string): string {
  return text.replace(
    INLINE_TAG,
    (match: string, leading: string, tagName: string, content: string) => {
      const split = splitLinkText(content);
      const target = split.target;
      const caption = leading || split.caption;

      return buildLink(match, target, caption, tagName === 'linkcode');
    },
  );
}

function buildLink(
  original: string,
  target: string,
  caption: string,
  monospace = false,
) {
  if (monospace) {
    caption = backTicks(caption);
  }

  if (URL_PREFIX.test(target)) {
    return link(caption, target);
  }

  const { project } = getContext();

  const reflection = project.findReflectionByName(target);
  if (reflection && reflection.url) {
    return linkTemplate(caption, reflection.url, !monospace);
  } else {
    return original;
  }
}

function splitLinkText(text: string) {
  let splitIndex = text.indexOf('|');
  if (splitIndex === -1) {
    splitIndex = text.search(/\s/);
  }
  if (splitIndex !== -1) {
    return {
      caption: text
        .substr(splitIndex + 1)
        .replace(/\n+/, ' ')
        .trim(),
      target: text.substr(0, splitIndex).trim(),
    };
  } else {
    return {
      caption: text,
      target: text,
    };
  }
}
