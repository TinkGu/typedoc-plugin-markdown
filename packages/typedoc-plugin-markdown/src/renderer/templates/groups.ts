import {
  DeclarationReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../tools/builder';
import { heading, horizontalRule } from '../tools/elements';
import { memberTemplate } from './member';

export function groupsTemplate(groups: ReflectionGroup[]) {
  const md = new MarkdownBuilder();

  groups
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      if (group.categories) {
        group.categories
          ?.filter((category) => !category.allChildrenHaveOwnDocument())
          .forEach((category) => {
            md.add(heading(2, `${group.title} ${category.title}`));
            md.add(getMembers(category));
          });
      } else {
        md.add(heading(2, group.title));
        md.add(getMembers(group));
      }
    });

  return md.print();
}

function getMembers(item: ReflectionGroup | ReflectionCategory) {
  const md = new MarkdownBuilder();
  item.children.forEach((child, index) => {
    md.add(memberTemplate(child as DeclarationReflection));
    if (index !== item.children.length - 1) {
      md.add(horizontalRule());
    }
  });
  return md.print();
}
