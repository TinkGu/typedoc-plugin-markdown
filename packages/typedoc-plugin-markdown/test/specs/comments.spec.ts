import settings from '../../src/renderer/settings';
import { commentsTemplate } from '../../src/renderer/templates/comments';
import { TestApp } from '../test-app';

describe(`Comments:`, () => {
  let testApp: TestApp;

  beforeAll(() => {
    testApp = new TestApp(['comments.ts']);
    testApp.bootstrap();
    settings.activeUrl = 'modules.md';
  });

  afterAll(() => {
    testApp.cleanup();
  });

  test(`should build @link references'`, () => {
    expect(
      commentsTemplate(testApp.findReflection('commentWithDocLinks').comment),
    ).toMatchSnapshot();
  });

  test(`should convert symbols brackets to symbol links'`, () => {
    expect(
      commentsTemplate(
        testApp.findReflection('commentsWithSymbolLinks').comment,
      ),
    ).toMatchSnapshot();
  });

  test(`should convert comments with fenced block'`, () => {
    expect(
      commentsTemplate(
        testApp.findReflection('commentsWithFencedBlock').comment,
      ),
    ).toMatchSnapshot();
  });

  test(`should convert comments with tags'`, () => {
    expect(
      commentsTemplate(testApp.findReflection('commentsWithTags').comment),
    ).toMatchSnapshot();
  });

  test(`should allow html in comments'`, () => {
    expect(
      commentsTemplate(testApp.findReflection('commentsWithHTML').comment),
    ).toMatchSnapshot();
  });
});
