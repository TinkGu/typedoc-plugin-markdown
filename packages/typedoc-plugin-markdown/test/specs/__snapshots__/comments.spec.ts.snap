// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments: should allow html in comments' 1`] = `
"<p>
You can write <strong>HTML</strong> tags directly in comments
</p>
<ul>
<li>List item</li>
</ul>"
`;

exports[`Comments: should build @link references' 1`] = `
"See [\`CommentClass\`](classes/CommentClass.md) and [CommentClass's comment property](classes/CommentClass.md#comment).
Also, check out [Google](https://www.google.com) and
[GitHub](https://github.com).

Taken from [JsDoc](http://usejsdoc.org/tags-inline-link.html).
"
`;

exports[`Comments: should convert comments with fenced block' 1`] = `
"Some comments with fence blocks

\`\`\`typescript
someFunction()
\`\`\`

\`\`\`js
anotherFunction()
\`\`\`
"
`;

exports[`Comments: should convert comments with tags' 1`] = `
"**\`name\`** Tag description on same line

**\`description\`**
Tag description on new line

- Tag description on another line


**\`deprecated\`**
Another tag description
"
`;

exports[`Comments: should convert symbols brackets to symbol links' 1`] = `
"Additionally you can link to other classes, members or functions using double square brackets.

- Link to an external reflection: [CommentClass](classes/CommentClass.md)
- Link to an internal reflection: [commentsInReturn](modules.md#commentsinreturn)
- Link to an undefined reflection: [[VOID]]
"
`;
